auth/indieauth/client.go

```diff
@@ -0,0 +1,87 @@
+package indieauth
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/owncast/owncast/core/data"
+	"github.com/pkg/errors"
+)
+
+var pendingAuthRequests = make(map[string]*Request)
+
+// StartAuthFlow will begin the IndieAuth flow by generating an auth request.
+func StartAuthFlow(authHost, userID, accessToken, displayName string) (*url.URL, error) {
+	serverURL := data.GetServerURL()
+	if serverURL == "" {
+		return nil, errors.New("server URL must be set to use auth")
+	}
+
+	r, err := createAuthRequest(authHost, userID, displayName, accessToken, serverURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to generate IndieAuth request")
+	}
+
+	pendingAuthRequests[r.State] = r
+
+	return r.Redirect, nil
+}
+
+// HandleCallbackCode will handle the callback from the IndieAuth server
+// to continue the next step of the auth flow.
+func HandleCallbackCode(code, state string) (*Request, *Response, error) {
+	request, exists := pendingAuthRequests[state]
+	if !exists {
+		return nil, nil, errors.New("no auth requests pending")
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "authorization_code")
+	data.Set("code", code)
+	data.Set("client_id", request.ClientID)
+	data.Set("redirect_uri", request.Callback.String())
+	data.Set("code_verifier", request.CodeVerifier)
+
+	codeChallenge := createCodeChallenge(request.CodeVerifier)
+	if codeChallenge != request.CodeChallenge {
+		return nil, nil, errors.New("code challenge mismatch")
+	}
+
+	client := &http.Client{}
+	r, err := http.NewRequest("POST", request.Endpoint.String(), strings.NewReader(data.Encode())) // URL-encoded payload
+	if err != nil {
+		return nil, nil, err
+	}
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	r.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
+
+	res, err := client.Do(r)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer res.Body.Close()
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var response Response
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, nil, err
+	}
+
```

Comment:
This needs some more error handling in case it gets back an error from the IndieAuth server before continuing on to assume there is a successful response returned.

Reply:
Since every implementation is unique, is it safe to rely on http status codes to determine there was an error as opposed to errors in the return payload? There's no way to know what the error structure will look like.

Reply:
The error structure is OAuth 2's structure with "error" and "error_description" keys in the JSON response

Reply:
I didn't realize that, I'm certainly not using that structure in my IndieAuth response. I'll make sure to do that, but if I didn't know it's possible other implementations didn't know either.

Reply:
What should be returned as `error` vs `error_description`?

Reply:
Yeah we chose to link out to the OAuth spec instead of copy all the error responses into the IndieAuth spec. Here's the list: https://datatracker.ietf.org/doc/html/rfc6749#section-5.2